### DISCUSSION TITLE
how to ignore include path for /usr/include/linux

### DISCUSSION THREAD

**Discussion by yzr95924** [microsoft/vscode-cpptools#11442] [Q&A] [answered]
Hi all,

I am working on developing liunx kernel in vscode. When I include "<linux/module.h>" in the source file,  IntellisSense will jump to "/usr/include/linux/module.h" instead of "/usr/src/$(uname -r)/include/linux/module.h". Can anyone provide some hints to address this problem? Thanks.

**Comment by yzr95924** [comment]
I also try to add "/usr/src/$(uname -r)/include" in includePath field in c_cpp_properties.json，but it does not help.

**Reply by bobbrow** [reply]
Can you share your c_cpp_properties.json?  IncludePaths should be searched in order from first to last.

Another option to try is using the [Makefile Tools ](https://marketplace.visualstudio.com/items?itemName=ms-vscode.makefile-tools)extension which can manage your configurations for you.

**Reply by yzr95924** [reply]
Sure, actually I find this issue might be caused by the order of entries in includePath:

Originally, I configure the c_cpp_properties.json as follows:

```json
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${workspaceFolder}/**",
                "/usr/src/linux-headers-5.15.90.1-microsoft-standard-WSL2/arch/x86/include/generated",
                "/usr/src/linux-headers-5.15.90.1-microsoft-standard-WSL2/arch/x86/include",
                "/usr/src/linux-headers-5.15.90.1-microsoft-standard-WSL2/include"
            ],
            "defines": [
                "__GNUC__",
                "__KERNEL__=1",
                "MODULE=1"
            ],
            "compilerPath": "/usr/bin/clang",
            "intelliSenseMode": "linux-clang-x64",
            "compileCommands": "${workspaceFolder}/build/compile_commands.json",
            "cStandard": "c17",
            "cppStandard": "c++17"
        }
    ],
    "version": 4
}
```
it does not work. then I change to the one as follows:

```json
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "/usr/src/linux-headers-5.15.90.1-microsoft-standard-WSL2/arch/x86/include/generated",
                "/usr/src/linux-headers-5.15.90.1-microsoft-standard-WSL2/arch/x86/include",
                "/usr/src/linux-headers-5.15.90.1-microsoft-standard-WSL2/include",
                "${workspaceFolder}/**"
            ],
            "defines": [
                "__GNUC__",
                "__KERNEL__=1",
                "MODULE=1"
            ],
            "compilerPath": "/usr/bin/clang",
            "intelliSenseMode": "linux-clang-x64",
            "compileCommands": "${workspaceFolder}/build/compile_commands.json",
            "cStandard": "c17",
            "cppStandard": "c++17"
        }
    ],
    "version": 4
}
```
it works.

**Reply by bobbrow** [reply, marked as answer]
I'm a bit surprised since I thought the path that was interfering was `/usr/include` which comes from your compiler's system include path that is added after the `includePath` you've specified.  I'm wondering if your local version of `<linux/module.h>` has an `#include_next` in it?

The observation you've made above is correct though.  If you want a certain path to take precedence over the recursive search in your workspace folder (`${workspaceFolder}/**`), you add it at the beginning of the `includePath`.

**Comment by michelleangela** [comment]
@yzr95924 
The extension has these files exclude settings that you could use to exclude folders/files.

![image](https://github.com/microsoft/vscode-cpptools/assets/38734287/b3b0a890-0414-4865-a593-5454695e628c)

**Reply by yzr95924** [reply]
thanks, I will try this
